pkg/server/image_pull.go

```diff
@@ -17,14 +17,226 @@ limitations under the License.
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
 	"golang.org/x/net/context"
-
+	"golang.org/x/sync/errgroup"
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 )
 
 // PullImage pulls an image with authentication config.
+// TODO(mikebrow): add authentication
+// TODO(mikebrow): harden api (including figuring out at what layer we should be blocking on duplicate requests.)
 func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
-	return nil, errors.New("not implemented")
+	var (
+		err  error
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	image := r.GetImage().Image
+
+	if desc, size, err = c.pullImage(ctx, image); err != nil {
+		return nil, fmt.Errorf("failed to pull image %q: %v", image, err)
+	}
+	digest := desc.Digest.String() // TODO(mikebrow): add truncIndex for image id
+
+	// TODO(mikebrow): pass a metadata struct to pullimage and fill in the tags/digests
+	// store the image metadata
+	var meta *metadata.ImageMetadata
+	meta, err = c.imageMetadataStore.Get(digest)
+	if err == nil {
+		// TODO(mikebrow): consider what to do if pullimage was called and metadata already exists (error? udpate?)
+		if meta == nil {
+			meta = &metadata.ImageMetadata{
+				ID:          digest,
+				RepoTags:    []string{image},
+				RepoDigests: []string{digest},
+				Size:        uint64(size),          // TODO(mikebrow):  compressed or uncompressed size? using compressed
+				Auth:        r.GetAuth(),           // TODO(mikebrow): what to store here
+				CreatedAt:   time.Now().UnixNano(), // TODO(mikebrow): should this be the image created at time?
+			}
+			if err = c.imageMetadataStore.Create(*meta); err != nil {
+				return &runtime.PullImageResponse{ImageRef: digest},
+					fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and manifest via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ctx context.Context, ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
```

Comment:
Mentioned in previous comments that we only need `image.Config` now... This function may not be needed if it doesn't provide `image.Config`.

I'm fine with leave this function in this PR for now, but we may not need it in the future.